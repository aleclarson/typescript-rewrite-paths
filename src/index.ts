import MagicString, { SourceMap } from 'magic-string'
import * as ts from 'typescript'

export function rewritePaths(input: string, rewrite: (path: string) => string) {
  const output = new MagicString(input)
  extractPaths(input).forEach(i => {
    const text = rewrite(i.text)
    if (text !== i.text && typeof text == 'string') {
      output.overwrite(i.start, i.end, text)
    }
  })
  return output.toString()
}

export interface ImportPath {
  text: string
  start: number
  end: number
}

export function extractPaths(input: string) {
  const imports: ImportPath[] = []
  const module = ts.createSourceFile('', input, ts.ScriptTarget.Latest)
  module.forEachChild(node => {
    if (ts.isImportDeclaration(node)) {
      const origin = node.moduleSpecifier

      // Omit the quote characters.
      const start = origin.getStart(module) + 1
      const end = origin.getEnd() - 1

      imports.push({
        text: input.slice(start, end),
        start,
        end,
      })
    }
  })
  return imports
}
