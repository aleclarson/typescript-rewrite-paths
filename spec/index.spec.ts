import { rewritePaths, extractPaths } from '../src'
import dedent = require('dedent')

test('rewritePaths', () => {
  const code = dedent`
    import {a} from 'foo'
    import {a,b,c} from 'foo/lib/bar.js'
    import {a as b} from 'foo'
    import {a as b, c as d} from '@foo/bar'
    import * as foo from 'foo'
    import foo from 'foo'
    import foo, {a,b} from 'foo'
  `

  expect(extractPaths(code)).toMatchSnapshot()

  const result = rewritePaths(code, path => {
    // For fun, treat direct imports differently.
    if (path.endsWith('.js')) {
      return path.replace(/\.js$/, '.mjs')
    }
    // Skip scoped imports.
    if (path.startsWith('@')) {
      return path
    }
    // Prepend our scope to each import.
    return '@scope/' + path
  })

  expect(result).toMatchSnapshot()
})
