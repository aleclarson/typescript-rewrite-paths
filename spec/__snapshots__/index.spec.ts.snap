// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rewritePaths 1`] = `
Array [
  Object {
    "end": 20,
    "start": 17,
    "text": "foo",
  },
  Object {
    "end": 57,
    "start": 43,
    "text": "foo/lib/bar.js",
  },
  Object {
    "end": 84,
    "start": 81,
    "text": "foo",
  },
  Object {
    "end": 124,
    "start": 116,
    "text": "@foo/bar",
  },
  Object {
    "end": 151,
    "start": 148,
    "text": "foo",
  },
  Object {
    "end": 173,
    "start": 170,
    "text": "foo",
  },
  Object {
    "end": 202,
    "start": 199,
    "text": "foo",
  },
]
`;

exports[`rewritePaths 2`] = `
"import {a} from '@scope/foo'
import {a,b,c} from 'foo/lib/bar.mjs'
import {a as b} from '@scope/foo'
import {a as b, c as d} from '@foo/bar'
import * as foo from '@scope/foo'
import foo from '@scope/foo'
import foo, {a,b} from '@scope/foo'"
`;
